worker/index/index.go

```diff
@@ -0,0 +1,247 @@
+
+	mu := config.Lock().ReadWrite(inst, "index/"+msg.Doctype)
+	if err := mu.Lock(); err != nil {
+		return err
+	}
+	defer mu.Unlock()
+
+	lastSeq, err := getLastSeqNumber(inst, msg.Doctype)
+	if err != nil {
+		return err
+	}
+	feed, err := callChangesFeed(inst, msg.Doctype, lastSeq)
+	if err != nil {
+		return err
+	}
+	if feed.LastSeq == lastSeq {
+		return nil
+	}
+
+	for _, change := range feed.Results {
+		if err := callExternalIndexers(inst, msg.Doctype, change); err != nil {
+			logger.Warnf("Index error: %s", err)
+			return err
+		}
+	}
+	_ = updateLastSequenceNumber(inst, msg.Doctype, feed.LastSeq)
+
+	if feed.Pending > 0 {
+		_ = pushJob(inst, msg.Doctype)
+	}
+
+	return nil
+}
+
+func callExternalIndexers(inst *instance.Instance, doctype string, change couchdb.Change) error {
+	if strings.HasPrefix(change.DocID, "_design/") {
+		return nil
+	}
+	if change.Doc.Get("type") == consts.DirType {
+		return nil
+	}
+
+	indexers := inst.ExternalIndexers()
+	for _, indexer := range indexers {
+		u, err := url.Parse(indexer)
+		if err != nil {
+			return err
+		}
+		u.Path = fmt.Sprintf("/docs/%s/%s/%s", inst.Domain, doctype, change.DocID)
+		if change.Deleted {
+			req, err := http.NewRequest(http.MethodDelete, u.String(), nil)
+			if err != nil {
+				return err
+			}
+			res, err := http.DefaultClient.Do(req)
+			if err != nil {
+				return err
+			}
+			defer res.Body.Close()
+			if res.StatusCode >= 500 {
+				return fmt.Errorf("DELETE status code: %d", res.StatusCode)
+			}
+		} else {
+			md5sum := fmt.Sprintf("%x", change.Doc.Get("md5sum"))
+			req, err := http.NewRequest(http.MethodGet, u.String(), nil)
+			if err != nil {
+				return err
+			}
+			res, err := http.DefaultClient.Do(req)
+			if err != nil {
+				return err
+			}
+			defer res.Body.Close()
+
+			// When the content has not changed, there is no need to regenerate
+			// an embedding.
+			needIndexation := false
+			switch res.StatusCode {
+			case 200:
+				var metadata map[string]interface{}
+				if err = json.NewDecoder(res.Body).Decode(&metadata); err != nil {
+					return err
+				}
+				needIndexation = metadata["md5sum"] != md5sum
+			case 404:
+				needIndexation = true
+			default:
+				return fmt.Errorf("GET status code: %d", res.StatusCode)
+			}
+			if !needIndexation {
+				// TODO we should patch the metadata in the vector db when a
+				// file has been moved/renamed.
+				continue
+			}
+
+			dirID, _ := change.Doc.Get("dir_id").(string)
+			name, _ := change.Doc.Get("name").(string)
+			mime, _ := change.Doc.Get("mime").(string)
+			internalID, _ := change.Doc.Get("internal_vfs_id").(string)
+			u.RawQuery = url.Values{
+				"dir_id": []string{dirID},
+				"name":   []string{name},
+				"md5sum": []string{md5sum},
+			}.Encode()
+			fs := inst.VFS()
+			fileDoc := &vfs.FileDoc{
+				Type:       consts.FileType,
+				DocID:      change.DocID,
+				DirID:      dirID,
+				DocName:    name,
+				InternalID: internalID,
+			}
+			// TODO notes with images
```

Comment:
For now, we don't want to deal with images, right?

Reply:
Right, but currently, the file is not indexed at all (not even the text part). The stack sends the tarball, not the markdown, and the RAG just skip this file.

Reply:
Oh I see. And for notes without images, is it seen as raw text or actual markdown?

Reply:
Just raw text for the moment